### ISSUE TITLE
Figure out Graphitti starting sequence with loading graphs from GraphML files

### ISSUE THREAD

**Issue by jardiamj** [UWB-Biocomputing/Graphitti#377]
Moving into providing the initial Graph structure to Graphitti using GraphML files is forcing us to review its starting sequence.

- We must figure out where and when input and graph parameters must be loaded.
- We need to figure out if we can leverage the OperationsManager for some initialization steps that don't need to happen in a specific order.
- Finally, the naming of classes and methods must be reviewed.

This picture is a starting point.

![Graphitti_start_sequence](https://user-images.githubusercontent.com/1088732/200485288-c7d6d2e3-ac68-403a-bac7-554433041934.jpg)

**Comment by jardiamj** [comment]
In the image below, I traced some of the Class parameter dependencies in the current implementation.

# Observations about the initialization sequence

- Memory allocation happens in the `setup` method of each class (`setupLayout`, `setupVertiecs`, etc).
- The setup methods are called in `Model::setupSim()` in the following order:
- - setupVertices
- - setupEdges
- - setupLayout
- - createAllVertices --> This sets up the individual value for each vertex using the Layout's `vertexTypeMap_`
- - setupConnections
- - createEdgeIndexMap
- I believe that the order in which the setup methods are called doesn't matter.
- - The classes only need to know the number of vertices and/or maxEdges to allocate memory

# Observations about parameters dependencies

- The Simulator object (singleton) loads some parameters that are not needed until later in the initialization sequence such as:
- - totalNeurons (should be named numVertices), and
- - maxEdgesPerVertex
- These parameters are used by `Layout`, `Connections`, `AllVertices`, and `AllEdges`. (as described in the picture)
- maxEdgesPerVertex is only used by `AllEdges` and `Connections`. Since `Connections` manages `AllEdges` we could put it there.
- numVertices is used by `AllVertices`, `Layout`, `Connections`, and `AllEdges`. Both Layout and Connections need it, so it makes sense to be in Simulator or Model, although it can be set later on when the GraphML file is read.

![Graphitti_param_dependencies](https://user-images.githubusercontent.com/1088732/200687237-c6a14151-2ac4-4818-9094-757d4e5d1ec4.jpg)

**Comment by jardiamj** [comment]
I found out that the Recorder classes do allocate memory on initialization. We should probably move memory allocation to their `init()` method, which should probably be named `setupRecorder()` to keep the naming consistent. Once this is done, I believe we can delegate the setup step to the OperationsManager.

**Comment by stiber** [comment]
Since @PoojaPal2021 is rewriting the Recorder classes, @jardiamj just hack the existing ones to make this move to facilitate your work.

**Comment by jardiamj** [comment]
# Documenting changes

- Made the `GraphManager` a singleton since it no longer needs to be a template.
- Moved numVertices into Layout. The Simulator class retrieves it from there when asked about it.
- Moved the setting of X and Y locations out of the base class (`Layout`) into the subclass. Base only allocates memory for them.
- Moved the allocation of memory in `XmlGrowthRecorder` to the `init` methods. Not yet renamed.

# Notes:

- These changes decouple the Layout derived classes from the original grid structure.
- FixedLayout has been modified and tested.
- I still need to test DynamicLayout.

**Comment by jardiamj** [comment]
There are currently 2 things preventing us from moving the `setup` step to the operations manager:

1. When we add the edges in the Connections setup phase, we pass the summation point for the destination vertex into the `addEdge()` method. The memory allocation for the `summationMap_` is done in the `AllVertices` setup method, therefore it must happen first.
2. In the setup phase, `ConnGrowth` saves the distance from the dist_ matrix into a delta_ matrix, while ConnStatic uses it to gather neurons within a radius threshold. This distance is calculated in the `Layout` setup phase so it must also happen first.

If we separate the memory allocation from the creation of edges in the `Connections` classes we could delegate the `setup` phase to the `OperationsManager`.

**Comment by stiber** [comment]
So, let's think of decomposing these different "initialization/setup" actions into actions that aren't intermeshed in this way. We already have parameter reading separate. Rather than having a "setup" method (which is a rather generic and uninformative name, but which was reasonable when we didn't know exactly what each object would need to do), maybe these should be `allocMemory()` and [othername]. What is [othername]? `createContents()`? `initialize()`? That second one is almost as bad as `setup()` but maybe acceptable, since it's perhaps a catch basin for doing whatever the class needs to do after all parameters are loaded and all memory is allocated everywhere but before simulation commences.

**Comment by jardiamj** [comment]
`allocMemory()` and `initialize()` sound like reasonable names at the moment. I was planning to work on this in a PR separate from the one I have started already.

**Comment by jardiamj** [comment]
@stiber, I am wondering if this issue can be closed. I have completed the loading of GraphML files part and I think that @KamathScience has modified most of the other parts discussed here.

**Comment by stiber** [comment]
So, the question is: is this documented? I.e., is the current startup sequence reflected in the UML sequence diagrams? Is there a need to clarify anything (like in text documentation)?

**Comment by stiber** [comment]
Remaining things to do (each of these may require creating a separate issue):

- There are two top-level `.puml` files: `topLevelFlow.puml` and `graphitti.puml`. Need to compare them, produce a single, correct version, and get that version linked in and displaying in the GitHub pages.
- Double-check diagrams against the code. Do the diagrams show the startup that is actually implemented in master? Any changes here would be a hotfix.
- Assuming so, the next step would be to (after any PRs associated with the documentation of the master branch) merging those doc changes also into the development branch and repeating the double-check for the development branch. Any changes here would just be in development.
